### PR TITLE
feat(sdk): Python components - Parse component input/output descriptions from the function docstring

### PR DESCRIPTION
**Pull request by munagekar**
**Description of your changes:**

- improvements to docstring parsing

NOTE : I didn't find any changes in UI with this update, with kfp 1.0.0. This PR just lays the groundwork for #4513 , further changes are needed in sdk or backend to see any visible improvements.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by Ark-kun** [comment]
Thank you for this improvement.

File: sdk/python/kfp/components/_python_op.py [hunk 1]
```diff
@@ -258,11 +259,16 @@ def _capture_function_code_using_source_copy(func) -> str:
...
+def _extract_component_interface(func:Callable) -> ComponentSpec:
     single_output_name_const = 'Output'
 
     signature = inspect.signature(func)
     parameters = list(signature.parameters.values())
+
+    parsed_docstring = docstring_parser.parse(inspect.getdoc(func))
+    doc_dict = {p.arg_name: p.description for p in parsed_docstring.params}
+
+
```
Comment (Ark-kun) [inline sdk/python/kfp/components/_python_op.py:271]:
```suggestion
```

File: sdk/python/kfp/components/_python_op.py [hunk 2]
```diff
@@ -258,11 +259,16 @@ def _capture_function_code_using_source_copy(func) -> str:
     return func_code
 
 
-def _extract_component_interface(func) -> ComponentSpec:
+def _extract_component_interface(func:Callable) -> ComponentSpec:
```
Comment (Ark-kun) [inline sdk/python/kfp/components/_python_op.py:262]:
```suggestion
def _extract_component_interface(func: Callable) -> ComponentSpec:
```

File: sdk/python/kfp/components_tests/test_python_op.py [hunk 3]
```diff
@@ -454,6 +454,27 @@ def add_multiply_two_numbers(a: float = 3, b: float = 5) -> NamedTuple('DummyNam
...
+            """
+            Pipeline to Demonstrate Usage of Secret
+
+            Args:
+                env_var: Name of the variable inside the Pod
+                secret_name: Name of the Secret in the namespace
+            """
+
+        component_spec = comp._python_op._func_to_component_spec(pipeline)
+        self.assertEqual(component_spec.description,'Pipeline to Demonstrate Usage of Secret')
```
Comment (Ark-kun) [inline sdk/python/kfp/components_tests/test_python_op.py:473]:
```suggestion
        self.assertEqual(component_spec.description, 'Pipeline to Demonstrate Usage of Secret')
```

File: sdk/python/requirements.in [hunk 4]
```diff
@@ -4,6 +4,7 @@ PyYAML
 # kfp.components
 cloudpickle
 strip-hints>=0.1.8
+docstring-parser>=0.7.2
```
Comment (Ark-kun) [inline sdk/python/requirements.in:7]: Need to change requirements.txt as well

**Comment by munagekar** [comment]
@Ark-kun Thank you for the review. Sorry for the Lint Errors.

Regarding Lint, will you consider the usage of [black](https://github.com/psf/black) for python files, this is similar to gofmt. The line length is configurable. However, tab spacing is fixed to 4 spaces, this might lead to considerable amount of diff initially in some of the modules.

It seems docstring-parser doesn't support python<=3.5. Will you consider dropping support for python 3.5. Python 3.5 has reached [EOL in September 2020](https://www.python.org/dev/peps/pep-0478/). This will also allow the usage of field refs for type hints in the sdk. 

I will also try to add python 3.5 support to docstring-parser if it is trivial.

**Comment by munagekar** [comment]
Update: I am trying to get [py 3.5 support](https://github.com/rr-/docstring_parser/pull/28) on docstring-parser.

**Comment by Ark-kun** [comment]
> Will you consider dropping support for python 3.5.

Yes. I think we'll drop 3.5 soon.

**Comment by Ark-kun** [comment]
@munagekar I've discovered an issue with the docstring parser. https://github.com/rr-/docstring_parser/issues/29 If you have some time, can you please take a look?

**Comment by munagekar** [comment]
@Ark-kun I have sent in a [patch](https://github.com/rr-/docstring_parser/pull/31).

**Pull request merged**